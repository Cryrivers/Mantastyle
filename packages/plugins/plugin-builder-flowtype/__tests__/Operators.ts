import getTranspiledString from '../src/utils/transpile';

describe('Operators', () => {
  test('Unions', () => {
    expect(getTranspiledString('type Test = 1 | 2;')).toMatchSnapshot();
    expect(
      getTranspiledString('type Test<T,S> = T | S | 1 | 2;'),
    ).toMatchSnapshot();
  });
  test('Intersections', () => {
    expect(getTranspiledString('type Test = number & any;')).toMatchSnapshot();
    expect(
      getTranspiledString('type Test<T,S> = T & S & number;'),
    ).toMatchSnapshot();
  });
  test('Intersection and Union composed with brackets', () => {
    expect(
      getTranspiledString('type Test<T,S> = number & (1 | T | S)'),
    ).toMatchSnapshot();
  });
  test('$Keys', () => {
    expect(
      getTranspiledString(
        'type Keys = $Keys<{ a:number, b:string, c: boolean }>;',
      ),
    ).toMatchSnapshot();
    expect(
      getTranspiledString(
        'type Obj = { a:number, b:string, c: boolean }; type Keys = $Keys<Obj>;',
      ),
    ).toMatchSnapshot();
    expect(
      getTranspiledString('type Keys<T: Object> = $Keys<T>;'),
    ).toMatchSnapshot();
  });
  test('$Values', () => {
    expect(
      getTranspiledString(
        'type Values = $Values<{ a:number, b:string, c: boolean }>;',
      ),
    ).toMatchSnapshot();
    expect(
      getTranspiledString(
        'type Obj = { a:number, b:string, c: boolean }; type Values = $Values<Obj>;',
      ),
    ).toMatchSnapshot();
    expect(
      getTranspiledString('type Values<T: Object> = $Values<T>;'),
    ).toMatchSnapshot();
  });
  test('$PropertyType', () => {
    expect(
      getTranspiledString(
        "type Obj = { a:number, b:string, c: boolean }; type PropertyType = $PropertyType<Obj, 'b'>;",
      ),
    ).toMatchSnapshot();
  });
});
