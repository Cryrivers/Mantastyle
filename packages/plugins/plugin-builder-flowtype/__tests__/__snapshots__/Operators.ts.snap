// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operators $Keys 1`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Keys = _Runtime.TypeAliasDeclaration(\\"Keys\\", function (typeFactory) {
  const type = _Runtime.KeyOfKeyword(_Runtime.TypeLiteral(function (typeLiteral) {
    typeLiteral.property(\\"a\\", _Runtime.NumberKeyword, false, []);
    typeLiteral.property(\\"b\\", _Runtime.StringKeyword, false, []);
    typeLiteral.property(\\"c\\", _Runtime.BooleanKeyword, false, []);
  }));

  return type;
}, []);"
`;

exports[`Operators $Keys 2`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Obj = _Runtime.TypeAliasDeclaration(\\"Obj\\", function (typeFactory) {
  const type = _Runtime.TypeLiteral(function (typeLiteral) {
    typeLiteral.property(\\"a\\", _Runtime.NumberKeyword, false, []);
    typeLiteral.property(\\"b\\", _Runtime.StringKeyword, false, []);
    typeLiteral.property(\\"c\\", _Runtime.BooleanKeyword, false, []);
  });

  return type;
}, []);

const Keys = _Runtime.TypeAliasDeclaration(\\"Keys\\", function (typeFactory) {
  const type = _Runtime.KeyOfKeyword(Obj);

  return type;
}, []);"
`;

exports[`Operators $Keys 3`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Keys = _Runtime.TypeAliasDeclaration(\\"Keys\\", function (typeFactory) {
  const T = typeFactory.TypeParameter(\\"T\\", _Runtime.ObjectKeyword);

  const type = _Runtime.KeyOfKeyword(T);

  return type;
}, []);"
`;

exports[`Operators Intersection and Union composed with brackets 1`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Test = _Runtime.TypeAliasDeclaration(\\"Test\\", function (typeFactory) {
  const T = typeFactory.TypeParameter(\\"T\\");
  const S = typeFactory.TypeParameter(\\"S\\");

  const type = _Runtime.IntersectionType([_Runtime.NumberKeyword, _Runtime.UnionType([_Runtime.Literal(1), T, S])]);

  return type;
}, []);"
`;

exports[`Operators Intersections 1`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Test = _Runtime.TypeAliasDeclaration(\\"Test\\", function (typeFactory) {
  const type = _Runtime.IntersectionType([_Runtime.NumberKeyword, _Runtime.AnyKeyword]);

  return type;
}, []);"
`;

exports[`Operators Intersections 2`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Test = _Runtime.TypeAliasDeclaration(\\"Test\\", function (typeFactory) {
  const T = typeFactory.TypeParameter(\\"T\\");
  const S = typeFactory.TypeParameter(\\"S\\");

  const type = _Runtime.IntersectionType([T, S, _Runtime.NumberKeyword]);

  return type;
}, []);"
`;

exports[`Operators Unions 1`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Test = _Runtime.TypeAliasDeclaration(\\"Test\\", function (typeFactory) {
  const type = _Runtime.UnionType([_Runtime.Literal(1), _Runtime.Literal(2)]);

  return type;
}, []);"
`;

exports[`Operators Unions 2`] = `
"import _Runtime from \\"@manta-style/runtime\\";

const Test = _Runtime.TypeAliasDeclaration(\\"Test\\", function (typeFactory) {
  const T = typeFactory.TypeParameter(\\"T\\");
  const S = typeFactory.TypeParameter(\\"S\\");

  const type = _Runtime.UnionType([T, S, _Runtime.Literal(1), _Runtime.Literal(2)]);

  return type;
}, []);"
`;
